### PR TITLE
feat(snap-rebuild): add io-engine controller api

### PR DESCRIPTION
**Pull request by tiagolobocastro**
Adds simple unit test to ensure api is working.
More will follow when the api is either exported via gRPC or through component/integration level testing.

### CODE REVIEW

File: control-plane/agents/src/bin/core/controller/io_engine/mod.rs [hunk 1]
```diff
@@ -220,3 +222,20 @@ pub(crate) trait HostApi {
     /// List blockdevices based on api versions.
     async fn list_blockdevices(&self, request: &GetBlockDevices) -> Result<BlockDevices, SvcError>;
 }
+
+/// Interface for snapshot rebuilding operations.
+#[async_trait]
+pub(crate) trait SnapshotRebuildApi {
```
Comment (dsharma-dc) [inline control-plane/agents/src/bin/core/controller/io_engine/mod.rs:228]: Would it be simpler to make these a part of ReplicaSnapshotApi? We're going to deal with a replica's snapshots at a time during snap rebuild so looks logically near.

**Review by Abhinandan-Purkait — APPROVED** [review]

File: control-plane/agents/src/bin/core/controller/io_engine/mod.rs (hunk 1 above)
Comment (tiagolobocastro) [reply control-plane/agents/src/bin/core/controller/io_engine/mod.rs:228]: In previous dataplane api PR we had decided to keep snapshot rebuild itself separate as it has not much in common with snapshot api.

File: control-plane/agents/src/bin/core/controller/io_engine/types.rs [hunk 2]
```diff
@@ -32,3 +32,51 @@ pub(crate) struct RebuildHistoryResp {
...
+/// List all or a specific snapshot rebuild.
+pub(crate) struct ListSnapRebuild {
+    /// Select a specific target replica.
+    pub(crate) uuid: Option<transport::ReplicaId>,
+}
+/// A list of snapshot rebuilds.
+pub(crate) type ListSnapRebuildRsp = Vec<SnapshotRebuild>;
+/// Destroy a snapshot rebuild.
+pub(crate) struct DestroySnapRebuild {
+    /// The uuid of the particular rebuild to destroy.
```
Comment (dsharma-dc) [inline control-plane/agents/src/bin/core/controller/io_engine/types.rs]:
```suggestion
    /// The uuid of the particular replica for which to destroy the rebuild.
```

**Review by dsharma-dc — APPROVED** [review]

**Review by abhilashshetty04 — APPROVED** [review]

**Comment by abhilashshetty04** [comment]
@tiagolobocastro , should we call it TerminateSnapRebuild instead of DestroySnapRebuild ?

File: control-plane/agents/src/bin/core/controller/io_engine/types.rs (hunk 2 above)
Comment (tiagolobocastro) [reply control-plane/agents/src/bin/core/controller/io_engine/types.rs]: Ah though this is also the uuid of  the rebuild itself, which inherits its uuid from the replica. I'll also add this info in the doc comment

**Comment by tiagolobocastro** [comment]
> @tiagolobocastro , should we call it TerminateSnapRebuild instead of DestroySnapRebuild ?

On the dataplane api it's destroy, so let's keep the same terminology

**Comment by tiagolobocastro** [comment]
bors merge

**Pull request merged**